### PR TITLE
Convert Parent Id to Context before calling the ActivityListener

### PR DESCRIPTION
**Pull request by tarekgh**
When API users call `ActivitySource.StartActivity(..., string parentId,....) `, we eventually call all listeners callback `ActivityListener.GetRequestedDataUsingParentId` and we pass the parent Id string to that callback. parent Id string can be a regular string or it can be constructed from the parent context (trace Id, span Id, and the trace flags). The feedback we got from OT is this design will require the implementers to try to extract the context from the parent Id. If the context cannot be extracted, the passed information will be not useful for the OT scenarios (the parent Id can be useful in other scenarios though). That is mean the burden will be on the listener implementer and will require them to implement `ActivityListener.GetRequestedDataUsingParentId` while not be used in some cases when cannot extract the context. 

The proposed change here is when calling `ActivitySource.StartActivity(..., string parentId,....)` and the listener implementing the callback `ActivityListener.GetRequestedDataUsingParentId`, we'll call this callback as we do today as the listener showing the intent to receive and handle the Parent Id. If the listener didn't implement  `ActivityListener.GetRequestedDataUsingParentId` and implementing `ActivityListener.GetRequestedDataUsingContex` then we'll try to convert the parent Id to context. If the conversion succeeds well call `ActivityListener.GetRequestedDataUsingContex` otherwise we'll not call any listeners callbacks.
There is no APIs change here as all design change is more on the internal implementation details.

### CODE REVIEW

**Comment by tarekgh** [comment]
CC @cijothomas @reyang

File: src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/ActivitySource.cs [hunk 1]
```diff
@@ -117,7 +117,7 @@ public bool HasListeners()
 
             ActivityDataRequest dataRequest = ActivityDataRequest.None;
 
-            if (parentId != null)
+            if (parentId != null && !Activity.TryConvertIdToContext(parentId, out context))
```
Comment (noahfalk) [inline src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/ActivitySource.cs:120]: I would suggest only converting the id as a fallback if the GetDataRequestByParentId callback wasn't provided. This lets people write optimized implementations of a sampler that skip the CPU and allocation costs of conversion if they want to. I am aiming that it should be possible for people to write optimized samplers that operate in double-digit ns. Based on benchmarking from the other day I'd expect this call alone to be at least 50ns, and probably a good bit more.

**Review by noahfalk — APPROVED** [review]
One comment inline, otherwise looks good to me!

File: src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/ActivitySource.cs (hunk 1 above)
Comment (tarekgh) [reply src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/ActivitySource.cs:120]: @cijothomas are you ok with what @noahfalk suggesting here? I assume it should be ok but wanted to confirm before proceeding applying the suggestion.

Comment (cijothomas) [reply src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/ActivitySource.cs:120]: Yes. this sounds better.

**Comment by tarekgh** [comment]
@noahfalk I have addressed your feedback. Please let me know if this matches what you have suggested.

File: src/libraries/System.Diagnostics.DiagnosticSource/tests/ActivitySourceTests.cs [hunk 2]
```diff
@@ -380,6 +380,56 @@ public void TestDefaultParentContext()
...
+
+                using ActivitySource aSource = new ActivitySource("ParentIdsTest");
+                using ActivityListener listener1 = new ActivityListener();  // will have context callback only
+                using ActivityListener listener2 = new ActivityListener();  // will have parent id  callback only
+                using ActivityListener listener3 = new ActivityListener();  // will have both context and parent Id callbacks
+
+                listener1.ShouldListenTo = listener2.ShouldListenTo = listener3.ShouldListenTo = (activitySource) => activitySource.Name == "ParentIdsTest";
+                listener1.GetRequestedDataUsingContext = listener3.GetRequestedDataUsingContext = (ref ActivityCreationOptions<ActivityContext> activityOptions) =>
+                {
+                    callingByContext++;
```
Comment (cijothomas) [inline src/libraries/System.Diagnostics.DiagnosticSource/tests/ActivitySourceTests.cs:400]: Do you also need to validate the the activityOptions.Parent.Context is matching the passed parentid - const string w3cId = "00-99d43cb30a4cdb4fbeee3a19c29201b0-e82825765f051b47-00";

Comment (tarekgh) [reply src/libraries/System.Diagnostics.DiagnosticSource/tests/ActivitySourceTests.cs:400]: good idea. I'll do that.

**Comment by noahfalk** [comment]
> Please let me know if this matches what you have suggested.

Yep! Looks good : )

**Review by cijothomas — APPROVED** [review]

**Pull request merged**